internal/app/pipeline/process/process.go

```diff
@@ -51,6 +51,9 @@ type Service interface {
 
 	// CancelProcess cancels a single process.
 	CancelProcess(ctx context.Context, id string) (err error)
+
+	// SignalProcess sends a signal to a single process.
+	SignalProcess(ctx context.Context, id string, signal string, value interface{}) (err error)
```

Comment:
I'm not sure this belongs here. The process log service is supposed to record these events. (In retrospect, `CancelProcess` doesn't belong here either, I think it was a temporary workaround for cancelling node pool upgrades.)

Reply:
now we have a separate process and process workflow interface. we don't use the former one anywhere though.

Reply:
Yeah, I don't think it's any better. Why does it have to be tied to the process service? The two are completely independent.

Reply:
because we use the process service for three things in the whole platform
1. to register processes from workflows
2. to list, and cancel them over the http api
3. to send signals to them based on status changes

processes that belong to a workflow are an abstraction of the workflow engine. if we want to do candence-specific things from other services with processes, then we wouldn't even need to use the processes, because we could query cadence. but that's not what we want to do

Reply:
TBH I like that even `CancelProcess` is here, it is very handy to use in my opinion here. If we wish to move them away to a different interface I would say we can do it later on if we think it makes more sense there. (This PR is quite BIG already).

Reply:
Adding `CancelProcess` was a convenience thing, so that the generic "process API" can be used to cancel processes, but it was never intended to be a permanent solution.

> processes that belong to a workflow are an abstraction for the workflow engine.

That's not entirely correct. The original idea behind the process **log** was to collect events that we can show the user as a log of what happened during the lifecycle of a cluster/nodepool/etc.

The process **log** API is supposed to have two interfaces actually:

- one exposed to the user for querying the log
- one exposed to workflows for writing the log

This concept introduces a third interface to the process **log**: workflow management.

I can accept merging it as is given that the next thing you do is refactoring this interface.

Reply:
What refactor would you actually be satisfied with? You want another service that can't list the processes, but only cancel and signal them?

Reply:
Yep, that would be a good start.

Please realize, that process **log** and process **management** are two completely different responsibilities.

It might not be obvious at first, because the current implementation does not reflect that properly (IMHO process log should only be written from processes, so the log content should travel in the signal, and the workflow should write the log once it received the signal)

The process log should only be written from workflows. Signaling _usually_ comes from the outside. The log is usually read by a consumer (UI/CLI). These are completely different responsibilities and should be handled as such. It's not about what I would be satisfied with.

Reply:
#3322